collectors/python.d.plugin/openldap/openldap.chart.py

```diff
@@ -0,0 +1,145 @@
+        'lines': [
+            ['entries_sent', None, 'incremental']
+        ]
+    },
+    'ldap_operations': {
+        'options': [None, 'Operations', 'bind_operations', 'ldap', 'monitorCounter', 'line'],
+        'lines': [
+            ['bind_operations', None, 'incremental'],
+            ['search_operations', None, 'incremental'],
+            ['unbind_operations', None, 'incremental'],
+            ['add_operations', None, 'incremental'],
+            ['delete_operations', None, 'incremental'],
+            ['modify_operations', None, 'incremental'],
+            ['compare_operations', None, 'incremental']
+        ]
+    },
+    'waiters': {
+        'options': [None, 'Waiters', 'amount', 'ldap', 'monitorCounter', 'line'],
+        'lines': [
+            ['write_waiters', None, 'incremental'],
+            ['read_waiters', None, 'incremental']
+        ]
+    },
+}
+
+# Stuff to gather - make tuples of DN dn and attrib to get
+searchlist = {
+    'total_connections':('cn=Total,cn=Connections,cn=Monitor', 'monitorCounter'),
+    'bytes_sent': ('cn=Bytes,cn=Statistics,cn=Monitor', 'monitorCounter'),
+    'completed_operations': ('cn=Operations,cn=Monitor', 'monitorOpCompleted'),
+    'initiated_operations': ('cn=Operations,cn=Monitor', 'monitorOpInitiated'),
+    'referrals_sent': ('cn=Referrals,cn=Statistics,cn=Monitor', 'monitorCounter'),
+    'entries_sent': ('cn=Entries,cn=Statistics,cn=Monitor', 'monitorCounter'),
+    'bind_operations': ('cn=Bind,cn=Operations,cn=Monitor', 'monitorOpCompleted'),
+    'unbind_operations': ('cn=Unbind,cn=Operations,cn=Monitor', 'monitorOpCompleted'),
+    'add_operations': ('cn=Add,cn=Operations,cn=Monitor', 'monitorOpInitiated'),
+    'delete_operations':  ('cn=Delete,cn=Operations,cn=Monitor', 'monitorOpCompleted'),
+    'modify_operations': ('cn=Modify,cn=Operations,cn=Monitor', 'monitorOpCompleted'),
+    'compare_operations': ('cn=Compare,cn=Operations,cn=Monitor', 'monitorOpCompleted'),
+    'search_operations': ('cn=Search,cn=Operations,cn=Monitor', 'monitorOpCompleted'),
+    'write_waiters': ('cn=Write,cn=Waiters,cn=Monitor', 'monitorCounter'),
+    'read_waiters': ('cn=Read,cn=Waiters,cn=Monitor', 'monitorCounter'),
+}
+
+class Service(SimpleService):
+    def __init__(self, configuration=None, name=None):
+        SimpleService.__init__(self, configuration=configuration, name=name)
+        self.order = ORDER
+        self.definitions = CHARTS
+
+        self.server = configuration.get('server')
+        self.port = configuration.get('port')
+        self.username = configuration.get('username')
+        self.password = configuration.get('password')
+        
+    def connect(self):
+        try: 
+            print self.server
+            self.conn = ldap.initialize('ldap://%s:%s' % ( self.server, self.port))
+            self.conn.simple_bind(self.username, self.password)
+            return True
+        except:
+            return False
```

Comment:
We need:
* connect
* reconnect
* alive flag

Thats's example. I haven't tested it. Feel free to use and modify.
<details><summary>click me!</summary>
<p>
```python
    def __init__(self, configuration=None, name=None):
        SimpleService.__init__(self, configuration=configuration, name=name)
        self.order = ORDER
        self.definitions = CHARTS

        self.server = configuration.get('server', 'localhost')
        self.port = configuration.get('port', 389)
        self.user = configuration.get('username')
        self.password = configuration.get('password')

        self.alive = False
        self.conn = None

    def disconnect(self):
        if self.conn:
            self.conn.unbind()
            self.conn = None
            self.alive = False

    def connect(self):
        self.conn = ldap.initialize('ldap://{0}:{1}'.format(self.server, self.port))

        try:
            self.conn.simple_bind(self.user, self.password)
        except ldap.LDAPError as error:
            self.error(error)
            return False

        self.alive = True
        return True

    def reconnect(self):
        self.disconnect()
        return self.connect()

    def check(self):
        if not HAS_LDAP:
            self.error("'python-ldap' package is needed")
            return None

        return self.connect() and self.get_data()

    def get_data(self):
        if not self.alive and not self.reconenct():
            return None
       #  other code
```

</p>
</details>

Reply:
to be honest, I don't realy understand when this can be false: 
... `and not self.reconnect():`
Where do we call this method?

Reply:
There is no guidelines, but this is natural to have _connect, reconnect_ methods for modules which connect to something (module should survive monitored service restart, stop/start). 

>  don't really understand when this can be false ...and not self.reconnect()

```python
if not self.alive and not self.reconenct():
    return None
```

This is always `False` while `alive` flag is `True`.
Module checks `alive` flag and if it's `True` it skips `reconnect` part.
If flag is `False` it calls `reconnect`.

Change `alive` flag to `False` if there is any `ldap` error occurs. This is neede for module to survive monitored service restart, stop/start